language/proto/gen/update_proto_csv_from_query.go

```diff
@@ -0,0 +1,212 @@
+// +build ignore
+
+/* Copyright 2019 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//
+// This code processes googleapis.xml xml file generated by the following Bazel query:
+//     bazel query 'kind("proto_library rule", //google/...)' --output xml --experimental_enable_repo_mapping > googleapis.xml
+//
+
+package main
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strings"
+)
+
+type Query struct {
+	XMLName xml.Name `xml:"query"`
+	Version string   `xml:"version,attr"`
+	Rules   []Rule   `xml:"rule"`
+}
+
+type Rule struct {
+	XMLName xml.Name `xml:"rule"`
+	Class   string   `xml:"class,attr"`
+	Name    string   `xml:"name,attr"`
+	Strings []String `xml:"string"`
+	Labels  []Label  `xml:"label"`
+	Lists   []List   `xml:"list"`
+}
+
+type List struct {
+	XMLName xml.Name `xml:"list"`
+	Name    string   `xml:"name,attr"`
+	Labels  []Label  `xml:"label"`
+}
+
+type Label struct {
+	XMLName xml.Name `xml:"label"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+type String struct {
+	XMLName xml.Name `xml:"string"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+func processQueryXml(xmlPath, repoPrefix string, protoContent *bytes.Buffer) (err error) {
+	query, err := readQueryXml(xmlPath)
+	if err != nil {
+		return
+	}
+
+	relPathList, err := processQuery(query, repoPrefix)
+	for _, v := range relPathList {
+		if _, err = fmt.Fprintf(protoContent, "%s\n", v); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func readQueryXml(xmlPath string) (query Query, err error) {
+	xmlFile, err := os.Open(xmlPath)
+	if err != nil {
+		fmt.Println(err)
+		return query, err
+	}
+	defer xmlFile.Close()
+
+	byteValue, err := ioutil.ReadAll(xmlFile)
+	if err != nil {
+		return query, err
+	}
+
+	// golang cannot process xml version > 1.0 yet. It seems like the generated xml is still a
+	// valid 1.0 one, so the following hack is "Ok".
+	xmlVersion := byteValue[:19]
+	if string(xmlVersion) == `<?xml version="1.1"` {
+		copy(xmlVersion, `<?xml version="1.0"`)
+	}
+
+	err = xml.Unmarshal(byteValue, &query)
+	return query, err
+}
+
+func processQuery(query Query, repoPrefix string) (relPathList []string, err error) {
```

Comment:
I'm a bit unsure of the format of the data in `relPathList` and `protoLabelMap`.

The final table should have one row per .proto file. So I think you need two maps:

* A map from .proto source files to labels of the `proto_library` targets that import them.
* A map from labels of `proto_library` targets to the labels of the `go_proto_library` targets that depend on them.

Those maps can be constructed by separate functions, independently.

Once you have them, you can build a list of keys from the first map, sort it, then print each row of the final table (effectively inner-joining the maps on proto label).

Reply:
I don't think we need a second map here. We can construct both, but after we constructed the second map, the only thing we would do with it is to iterate over it. Thus no need to construct the second map, we can just process the items of the "second map" on the fly, without storing them.

The generated final table has one row per `.proto` file. The only way it can have multiple rows per proto file if there is more than one `go_proto_library` rule declared for that file. Note, we do not control how the targets are created, so some of the older assumptions do not apply anymore. I actually do not know what to do in this case, as I think it is a non-supported scenario for gazelle as of now. It is also very unlikely to happen to be honest. The code above for such case will output multiple rows per same proto file (because it had more than one `go_proto_library` declared for it, so it would be a correct representation of what's inside the googleapis repo).

The generation strategy assumes that we need to handle only those protos, which have both targets declared for them `proto_library` and, transitively `go_proto_library`.  Thus, if you try to run this command against the current googleapis repo, it will output empty protos.csv (no `go_proto_library` rules are there yet).

The generation algorithm does the following: 
1. On first pass, process all `go_proto_library` rules. A `.proto` has a chance to be mentioned in `proto.csv` only in case if it has a corresponding `go_proto_library`. The protos without `go_proto_library` rules will be ignored. During this pass we are able to construct three of the four components of each line in proto.csv (2nd, 3rd and 4th). 
    1. 2nd - `proto` (or one for each element in `protos`) argument value (this is also the key in the constructed map)
    2. 3rd - `importpath` argument value
    3. 4th - the name of this `go_proto_library`

    The only missing piece is the path to the actual .proto file (1st component, the only "unique" in items in the `proto.csv` file). 

2. To find the path to the .proto files, on second pass process all `srcs` arguments of all `proto_library` targets and and map them to a corresponding `go_proto_library` targets (by using the map, constructed on step 1). If no corresponding `go_proto_library` target was found, ignore the .proto file, otherwise use the value of the `srcs` item as the 1st element in proto.csv row. At this point we have all 4 components of proto.csv row and ready to output. 
3) At the end of everythin we also sort the result (not sure if it is necessary, but it makes proto.csv look more like it was created from file tree traversal and more organized in general).

Reply:
This general approach sounds fine to me. One question though: will there be any `proto_library` targets in the wfuture (after all the rules are in place) that don't have corresponding `go_proto_library` targets but need to be available as dependencies for outside `proto_libraries`? If so, we may want to consider a more flexible approach (but it doesn't need to be in this PR).

Reply:
I'm not sure what you mean by the outside `proto_libraries`, but I want to say "No" (in general I think it is a pretty safe approach). All the core libraries will have `go_proto_library` defined for them, as well as all API-specific protos which choose to have Golang gapic client (which should be ALL apis, except the cases when the lib configuration is till "in progress").